File: lib/codeintel/autoindex/inference/clang.go
package inference

import (
	"path/filepath"
	"strings"

	"github.com/grafana/regexp"

	"github.com/sourcegraph/sourcegraph/lib/codeintel/autoindex/config"
)

func InferClangIndexJobHints(gitserver GitClient, paths []string) (hints []config.IndexJobHint) {
	inferredDir := make(map[string]bool)

	for _, path := range paths {
		dir := filepath.Dir(path)
		if inferredDir[dir] {
			continue
		}
		if strings.ToLower(filepath.Base(path)) == "cmakelists.txt" {
			hints = append(hints, config.IndexJobHint{
				Root:           dir,
				Indexer:        "sourcegraph/lsif-clang",
				HintConfidence: config.HintConfidenceProjectStructureSupported,
			})
			inferredDir[dir] = true
			continue
		}

		ext := strings.ToLower(filepath.Ext(path))
		if ext == ".cpp" || ext == ".c" || ext == ".h" || ext == ".hpp" || ext == ".cxx" || ext == ".cc" {
			hints = append(hints, config.IndexJobHint{
				Root:           dir,
				Indexer:        "sourcegraph/lsif-clang",
				HintConfidence: config.HintConfidenceLanguageSupport,
			})
			inferredDir[dir] = true
		}
	}

	return
}

func ClangPatterns() []*regexp.Regexp {
	return []*regexp.Regexp{
		suffixPattern(pathPattern(rawPattern("[cC][mM]ake[lL]ists.txt"))),
		extensionPattern(rawPattern("cpp")),
		extensionPattern(rawPattern("c")),
		extensionPattern(rawPattern("h")),
		extensionPattern(rawPattern("hpp")),
		extensionPattern(rawPattern("cxx")),
		extensionPattern(rawPattern("cc")),
	}
}

File: internal/codeintel/autoindexing/internal/inference/lua/clang.lua
local fun = require "fun"
local path = require "path"
local pattern = require "sg.autoindex.patterns"
local recognizer = require "sg.autoindex.recognizer"

local indexer = "sourcegraph/lsif-clang"

local is_cmakelist_file = function(base)
  return string.lower(base) == "cmakelists.txt"
end

return recognizer.new_path_recognizer {
  patterns = {
    pattern.new_path_extension_set { "c", "cc", "cpp", "cxx", "h", "hpp" },
    pattern.new_path_basename "CMakeLists.txt",
    pattern.new_path_basename "CMakelists.txt",
    pattern.new_path_basename "CmakeLists.txt",
    pattern.new_path_basename "Cmakelists.txt",
    pattern.new_path_basename "cMakeLists.txt",
    pattern.new_path_basename "cMakelists.txt",
    pattern.new_path_basename "cmakeLists.txt",
    pattern.new_path_basename "cmakelists.txt",
  },

  -- Invoked when c, cpp, header, or cmakelist files exist
  hints = function(_, paths)
    local hints = {}
    local visited = {}

    fun.each(function(p)
      local dir, base = path.split(p)

      if visited[dir] == nil and is_cmakelist_file(base) then
        table.insert(hints, {
          root = dir,
          indexer = indexer,
          confidence = "PROJECT_STRUCTURE_SUPPORTED",
        })

        visited[dir] = true
      end
    end, paths)

    fun.each(function(p)
      local dir, base = path.split(p)

      if visited[dir] == nil and not is_cmakelist_file(base) then
        table.insert(hints, {
          root = dir,
          indexer = indexer,
          confidence = "LANGUAGE_SUPPORTED",
        })

        visited[dir] = true
      end
    end, paths)

    return hints
  end,
}
